event-sourcing/event-source/src/main/java/uk/gov/justice/services/eventsourcing/source/core/DataSourceProducer.java

```
@@ -0,0 +1,58 @@
+package uk.gov.justice.services.eventsourcing.source.core;
+
+import uk.gov.justice.services.eventsourcing.source.core.annotation.EventSourceName;
+import uk.gov.justice.subscription.domain.eventsource.Location;
+import uk.gov.justice.subscription.registry.EventSourceRegistry;
+
+import java.util.Optional;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.InjectionPoint;
+import javax.inject.Inject;
+import javax.naming.NamingException;
+import javax.sql.DataSource;
+
+@ApplicationScoped
+public class DataSourceProducer {
+
+    @Inject
+    InitialContextProducer initialContextProducer;
+
+    @Inject
+    EventSourceNameExtractor eventSourceNameExtractor;
+
+    @Inject
+    EventSourceRegistry eventSourceRegistry;
+
+    @Produces
+    @EventSourceName
+    public DataSource dataSource(final InjectionPoint injectionPoint) {
+
+        final String eventSourceName = eventSourceNameExtractor.getEventSourceNameFromQualifier(injectionPoint);
+
+        final Optional<uk.gov.justice.subscription.domain.eventsource.EventSource> eventSourceFor = eventSourceRegistry.getEventSourceFor(eventSourceName);
+
+        if (eventSourceFor.isPresent()) {
+
+            final Location location = eventSourceFor.get().getLocation();
+
+            if (location.getDataSource().isPresent()) {
+                final Optional<String> dataSourceLocation = location.getDataSource();
+                return createDataSource(eventSourceName, dataSourceLocation);
+            } else {
+                return null;
```

I'd be tempted to throw an Exception here rather than returning null